File: correli.f
      parameter(nmax=4096)
      DIMENSION sismo1(4096),sismo2(4096),res(4096),x(4096),
     $          vec1(4096),vec2(4096)
      complex cor(8192)
      integer sizear
      character*100 arch1,arch2,outpfile,outpfile1,arch1n,arch2n
      write(*,*)'File no 1'
      read(*,*)arch1
      write(*,*)'File no 2'
      read(*,*)arch2
      sizear=lnblnk(arch1)
      arch1n=arch1(1:sizear-3)//'dat'
      write(*,*)arch1n
      write(*,*)sizear
      sizear=lnblnk(arch2)
      arch2n=arch2(1:sizear-3)//'dat'
      write(*,*)arch2n
      write(*,*)sizear
      open (16,file=arch1n)
      open (17,file=arch2n)

c     arch1='uno.1.sac'
c     arch2='cuatro.1.sac'
      ventana=0.3
      call RSAC1(arch1,sismo1,npts1,beg,delta,4096,nerr)
      if (nerr.ne.0) then
          write(*,*)'Some problem reading file ',arch1
          write(*,*)nerr
          stop
      endif
      call RSAC1(arch2,sismo2,npts2,beg,delta,4096,nerr)
      if (nerr.ne.0) then
          write(*,*)'Some problem reading file ',arch2
          stop
      endif
c     Renaming files
      outpfile = 'CORR.'//arch1(45:46)//'-'//arch2(45:46)//'.'
     $           //arch1(5:15)//'.dat'
      write(*,*)'Is the following name ok?, (if yes just press enter)'
      write(*,*)outpfile
      read(*,'(a)')outpfile1
      if(outpfile1(1:1).eq.'')outpfile1=outpfile
      open(15,file = outpfile1)
      limite1=1
      limite2=int(ventana/delta)
      nptsmax=npts1
      if(npts1.ge.npts2)nptsmax=npts2
      ancdelta=0.0
      aintervalo=0.0
      do j=1,4000
         if (limite1.gt.nptsmax.or.limite2.gt.nptsmax)goto 1999
c        Define time window
         aintervalo=aintervalo+ventana
         limite2=limite2+1
         k=0
         do i=1,nptsmax
            vec1(i)=0.0
            vec2(i)=0.0
         enddo
         aint1=0.0
         aint2=0.0
         do i=limite1,limite2
            k=k+1
            vec1(k)=sismo1(i)
            aint1=aint1+abs(sismo1(i))**2
            vec2(k)=sismo2(i)
            aint2=aint2+abs(sismo2(i))**2
         enddo
c        limite1=limite1+1
          
         if (npts1.gt.4096)then
            write(*,*)'exiting because number of points'
            goto 1999
         endif
         if (npts1.le.4096.or.npts2.le.1024)npts=4096
         if (npts1.le.2048.or.npts2.le.1024)npts=2048
         if (npts1.le.1024.or.npts2.le.1024)npts=1024
         if (npts1.le.512.or.npts2.le.512)npts=512
         if (npts1.le.256.or.npts2.le.256)npts=256
         if (npts1.le.128.or.npts2.le.128)npts=128
         if (npts1.le.64.or.npts2.le.64)npts=64
         do i=1,npts
            res(i)=0.0
         enddo
          
         call correl(vec2,vec1,npts,cor)

         k=(npts/4)+1
         do i=1,npts/4
            res(k)=real(cor(i))
            k=k+1
         enddo
         k=(npts/4)+1
c        write(*,*)k
         do i=1,npts/4
            res(i)=real(cor(k))
            k=k+1
         enddo
         corrim=-(delta*npts)/2.0
         delta1=delta*2.0
*        write(*,*)limite1,limite2,corrim,delta1,npts/4,nptsmax,
*    $             npts1,npts2
c        call wsac1('salida.sac',res,npts/2,corrim,delta1,nerr)
c        call wsac1('sismo1.sac',vec1,npts/2,corrim,delta,nerr)
c        call wsac1('sismo2.sac',vec2,npts/2,corrim,delta,nerr)
c        Calculando la integral para normalizacion
         fac=(aint1*aint2)**0.5
c        Buscando el maximo
         amax=0.0000
         do i=1,nptsmax
              if(res(i).gt.amax)amax=res(i)
         enddo
         amaxc=amax/fac
c        Performing noise correction
         ll=0
         u1=0.0
         u2=0.0
         an1=0.
         an2=0.
         do i=limite1,limite2
            ll=ll+1
            u1=u1+vec1(i)**2
            u2=u1+vec2(i)**2
         enddo
            u1=u1/float(ll)
            u2=u2/float(ll)
         if (j.eq.1) then
            ll=0
            do i=limite1,limite2
               ll=ll+1
               an1=an1+vec1(i)**2
               an2=an1+vec2(i)**2
            enddo
            an1=an1/float(ll)
            an2=an2/float(ll)
         endif
         anoisec=(sqrt(1-(an1/u1))*sqrt(1-(an2/u2)))
         amaxc=amaxc/anoisec
         if (amaxc.gt.1.0)amaxc=1.0
c        Considerando lp 5 hz (promedio 3hz)
c         alpha=3897.0  Valores originales
c         beta=2250.0   Valores originales
         alpha=8313.0
         beta=4800.0
         sigma=2*(1-amaxc)/355.3
         anum=7*((2/alpha**6)+(3/beta**6))
         aden=(6/alpha**8)+(7/beta**8)
         dist=sqrt(sigma*anum/aden)
         write(15,15)float(limite2-limite1)*delta,amaxc,dist
         ancdelta=ancdelta+delta
         write(16,*)ancdelta,sismo1(j)
         write(17,*)ancdelta,sismo2(j)
      enddo
1999  continue
14    format(i4,6f8.3)
15    format(3f8.3)
      end

File: correli.py
import obspy
import numpy as np
import sys
import argparse

alpha = 8313.0
beta = 4800.0

def correl(data1, data2, N):
    fft1, fft2 = twofft(data1, data2, N)
    NO2 = N // 2
    fft2 - fft1*np.conj(fft2)/NO2
    fft2(0) = np.real(fft2(0)) + np.real(fft2(N/2+1))
    return realft(fft2, N/2, -1)
    

def realft(data, N, isign):
    theta = 3.141592653589793238 / N
    c1 = 0.5
    c2 = -0.5 * isign
    if isign == 1:
        four1(data, N, 1)
    else:
        theta = -theta
    wpr = -2.0 * np.sin(0.5 * theta)**2
    wpi = np.sin(theta)
    wr = 1.0 + wpr
    wi = wpi
    n2p3 = 2*N + 3
    for i in range(1, N//2+1, 1):
        i1 = 2*i - 1
        i2 = i1 + 1
        i3 = n2p3 - i2
        i4 = i3 + 1
        h1r = c1 * (data[i1] + data[i3])
        h1i = c1 * (data[i2] - data[i4])
        h2r = -c2 * (data[i2] + data[i4])
        h2i = c2 * (data[i1] - data[i3])
        data[i1] = h1r + wr*h2r - wi*h2i
        data[i2] = h1i + wr*h2i + wi*h2r
        data[i3] = h1r - wr*h2r + wi*h2i
        data[i4] = -h1i + wr*h2i + wi*h2r
        wtemp = wr
        wr = wr*wpr - wi*wpi + wr
        wi = wi*wpr + wtemp*wpi + wi
    if isign == 1:
        data[0] = (h1r := data[0]) + data[1]
        data[1] = h1r - data[1]

    else:
        data[0] = c1 * ((h1r := data[0]) + data[1])
        data[1] = c1 * (h1r - data[1])
        four1(data, N, -1)
    return data

def four1(data, nn, isign): 
    N=2**nn
    j=1
    for i in range(1, N+1, 2):
        if j > i:
            tempr = data[j]
            tempi = data[j+1]
            data[j], data[j+1] = data[i], data[i+1]
            data[i], data[i+1] = tempr, tempi
        m = N // 2
        while m >= 2 and j > m:
            j -= m
            m = m // 2
        j += m
    mmax = 2
    while N > mmax:
        istep = 2 * mmax
        theta = 2*np.pi / (isign * mmax)
        
        wpr = -2.0 * np.sin(0.5 * theta)
        wpi = np.sin(theta)
        wr = 1.0
        wi = 0.0
        for m in range(1, mmax, 2):
            for i in range(m, N+1, istep):
                j = i + mmax
                tempr = wr * data[j] - wi * data[j+1]
                tempi = wr * data[j+1] + wi * data[j]
                data[j] = data[i] - tempr
                data[j+1] = data[i+1] - tempi
                data[i] += tempr
                data[i+1] += tempi
            wr = (wtemp := wr) * wpr - wi * wpi + wr
            wi = wi * wpr + wtemp * wpi + wi
        mmax = istep

def twofft(data1, data2, fft1, fft2, N):
    c1 = 0.5  + 0.0j
    c2 = -0.0 + 0.5j

    fft1 = data1 + data2*1j
    fft1 = four1(fft1, N, 1)
    fft2(0) = np.imag(fft1(0)) + 0j
    fft1(0) = np.real(fft1(0)) + 0j

    N2 = N + 2
    for j in range(2, N+1, 1):
        h1 = c1 * (fft1(j) + np.conj(fft1(N2-j)))
        h2 = c2 * (fft1(j) - np.conj(fft1(N2-j)))
        fft1(j) = h1
        fft1(N2-j) = np.conj(h1)
        fft2(j) = h2
        fft2(N2-j) = np.conj(h2)

    return fft1, fft2

if __name__ == "__main__":
    if len(sys.argv) != 4:
        parser = argparse.ArgumentParser(description="Process SAC files.")
        parser.add_argument("sac_file1", type=str, help="First SAC file")
        parser.add_argument("sac_file2", type=str, help="Second SAC file")
        parser.add_argument("output_file", type=str, help="Output file")
        parser.add_argument("-w", type=float, default=0.3, help="Window (default: 0.3s)")

        args = parser.parse_args()

        sac_file1 = args.sac_file1
        sac_file2 = args.sac_file2
        output_file = args.output_file
        window = args.w
        print("Usage: python correli.py <sac_file1> <sac_file2> <output_file>")
        sys.exit(1)


    try:
        st1 = obspy.read(sac_file1)
        st2 = obspy.read(sac_file2)
    except Exception as e:
        print(f"Error reading SAC files: {e}")
        sys.exit(1)

    delta1 = st1[0].stats.sac.delta
    delta2 = st2[0].stats.sac.delta
    npts1 = st1[0].stats.sac.npts
    npts2 = st2[0].stats.sac.npts

    if delta1 != delta2:
        print("Error: delta values are different")
        sys.exit(1)
    delta = delta1

    if npts1 != npts2:
        npts = min(npts1, npts2)

    lim_1 = 1
    lim_2 = np.round(window / delta)

    ancdelta = 0.0
    aintervalo = 0.0

    noise = True
    while True:
        aintervalo = aintervalo + window
        lim_2 = lim_2 + 1

        vec1 = st1[0].data[lim_1:lim_2]
        vec2 = st2[0].data[lim_1:lim_2]

        aint1 = np.sum(vec1**2)
        aint2 = np.sum(vec2**2)

        lim_1 = lim_1 + 1

        cor = correl(vec1, vec2, len(vec1))

        k = (npts/4) + 1
        res = np.zeros(npts/2)
        for i in range(0, npts/4, 1):
            res[k] = np.real(cor[i])
            k += 1
        
        for i in range(0, npts/4, 1):
            res[i] = np.real(cor[k])
            k += 1

        corrim = -(delta * npts/2)
        delta1 = delta*2
        fac = (aint1*aint2)**0.5

        amax = np.max(res)
        amaxc = amax/fac

        u1 = np.sum(vec1[lim_1:lim_2+1]**2)
        u2 = np.sum(vec2[lim_1:lim_2+1]**2)
        u1 = u1/(lim_2 - lim_1 + 1)
        u2 = u2/(lim_2 - lim_1 + 1)

        if noise:
            an1 = np.sum(vec1[lim_1:lim_2+1]**2)
            an2 = np.sum(vec2[lim_1:lim_2+1]**2)
            an1 = an1/(lim_2 - lim_1 + 1)
            an2 = an2/(lim_2 - lim_1 + 1)
            noise = False

        anoisec = np.sqrt(1 - (an1/u1))*np.sqrt(1 - (an2/u2))
        amaxc = amaxc/anoisec

        amax = np.clip(amax, None, 1)

        sigma = 2*(1 - amaxc)/355.3
        anum = 7*((2/alpha**6) + (3/beta**6))
        aden = (6/alpha**8) + (7/beta**8)
        dist = np.sqrt(sigma*anum/aden)
        ancdelta = ancdelta + delta

        













    

print("SAC files loaded successfully")
